README.md
<p align="center">
    <img src="Images/logos/hap-logo-rounded.png" width="33%">
</p>

# Home Access Plus+ for iOS

[Home Access Plus+ (HAP) for iOS provides a native app](https://itunes.apple.com/app/home-access-plus/id1076118300) to connect your Apple device to your institutions [Home Access Plus+](https://hap.codeplex.com) server. You can then browse, upload and download files easily to and from your iOS device to your institution network file drives. A [demo video](https://www.youtube.com/watch?v=cpr7Ar-a5MA) and [additional screenshots](https://goo.gl/photos/6Ptp4J1woojWuUik9) are available to view of the app in action (version 0.5.0).

<img src="Images/screenshots/screen-01.png" width="19%"><img src="Images/screenshots/spacer.gif" width="1%"><img src="Images/screenshots/screen-02.png" width="19%"><img src="Images/screenshots/spacer.gif" width="1%"><img src="Images/screenshots/screen-03.png" width="19%"><img src="Images/screenshots/spacer.gif" width="1%"><img src="Images/screenshots/screen-04.png" width="19%"><img src="Images/screenshots/spacer.gif" width="1%"><img src="Images/screenshots/screen-05.png" width="19%">

## Requirements
To be able to use this app, you will need to have the following:
* A device running iOS 9.0+
* [Home Access Plus+](https://hap.codeplex.com) set up and running for your institution (bug your network managers to get this set up&hellip; it's free)

## Frequently Asked Questions
Before asking for help or reporting a bug, please read through the [Frequently Asked Questions](FAQ.md) to see if the problem can be resolved.

## Contributing
Thanks for your interest in contributing to this project. You can contribute or report issues in the following ways:

### Beta Testing
* [Sign up for beta testing](http://www.edugeek.net/forums/home-access-plus/164699-home-access-plus-ios-app-request-testers.html) to explore and test the app through [Testflight](https://developer.apple.com/testflight/)
* [Download a release](https://github.com/stuajnht/HAP-for-iOS/releases) or [clone the latest](https://github.com/stuajnht/HAP-for-iOS.git) code and run it on your own device ([Xcode](https://developer.apple.com/xcode/) is needed to build the project)

### Reporting Issues
* [Report a bug](http://issuetemplate.com/#/stuajnht/HAP-for-iOS/bug) if something isn't working as you expect it to
* [Suggest an improvement](http://issuetemplate.com/#/stuajnht/HAP-for-iOS/request) for something that you'd like to see

### Pull Requests
If you would like to create a pull request, please make sure that you are on the [develop branch](https://github.com/stuajnht/HAP-for-iOS/tree/develop) before opening one. Once you have cloned or forked this repo, open the `HomeAccessPlus.xcworkspace` file in Xcode to begin development. This project uses [git-flow](https://github.com/nvie/gitflow) as its branching model.

The current development environment is with Xcode 8.2 and Swift 3.0.2. Please make sure you are using these versions before submitting any pull requests

### Contributors
The following users have contributed code or suggestions to this project:
* [@TestValleySchool](https://github.com/TestValleySchool):
  * [#19](https://github.com/stuajnht/HAP-for-iOS/pull/19)
* [kidpressingbuttons](http://www.edugeek.net/members/kidpressingbuttons.html)

## Privacy Purpose Permissions
Apps built for iOS 10+ must declare the purpose for access to any user private data types (sources: [1](http://useyourloaf.com/blog/privacy-settings-in-ios-10/), [2](https://developer.apple.com/videos/play/wwdc2016-709/?time=1470), [3](https://developer.apple.com/videos/play/wwdc2016-709/?time=1780)). Since this project uses the [PermissionScope Cocoapod](#cocoapods) a number of additional unused permissions are requested ([GitHub Issue](https://github.com/nickoneill/PermissionScope/issues/194)). Please see the [Frequently Asked Questions](FAQ.md#why-are-so-many-privacy-purpose-permissions-needed) for more information.

## License Terms
Home Access Plus+ for iOS is publised under the GNU GPL v3 License, see the [LICENSE](LICENSE.md) file for more information.

The name "Home Access Plus+" and all server side code are copyright Nick Brown ([nb development](https://nbdev.uk/projects/hap.aspx))

### Cocoapods
This project uses Cocoapods. Their project source code pages and licenses can be found below:
* [Alamofire](https://github.com/Alamofire/Alamofire/)
* [ChameleonFramework](https://github.com/ViccAlexander/Chameleon)
* [DKImagePickerController](https://github.com/zhangao0086/DKImagePickerController)
* [Font Awesome Swift](https://github.com/Vaberer/Font-Awesome-Swift)
* [Locksmith](https://github.com/matthewpalmer/Locksmith)
* [MBProgressHUD](https://github.com/jdg/MBProgressHUD)
* [PermissionScope](https://github.com/nickoneill/PermissionScope)
* [SwiftyJSON](https://github.com/SwiftyJSON/SwiftyJSON)
* [XCGLogger](https://github.com/DaveWoodCom/XCGLogger)
* [Zip](https://github.com/marmelroy/Zip)

### Other Projects
The following projects and source code are included in HAP+ for iOS. Their licenses project pages can be found below:
* [Reach](https://github.com/Isuru-Nanayakkara/Reach)

## To-Do List
The following features are planned for the Home Access Plus+ iOS app, along with their expected releases (which can change).

### 1.1.0
* Allow the app to be used with multiple Home Access Plus+ servers

### 1.2.0
* "Favourite" folders for devices in personal or single mode, using a tab bar to select between the my files or favourites "views" -- the use of swiping the table row right to select items could be used to show a favourite button
* Create a zip file from any selected items items
* (Optional branding) Set a logo and theme (including app icon) colour based on those used by the school to use throughout the app -- information possibly collected by accessing a app-config.json file located on the HAP+ server somewhere
* Support for document provider open and move options for apps that do not support import and export -- possibly won't happen, as all attempts to get this working so far haven't been successful


CHANGES.md
# Home Access Plus+ iOS app change log

## 1.0.1

### Changes
* Included additional information in the login view alert to inform the user to check the server is working, has a valid SSL certificate, is running TLS 1.2 and is compliant with App Transport Security requirements

## 1.0.0

### New Stuff
* "Upload" and "Create" options in the upload popover are disabled if the user does not have write access granted for the current directory
* Items can be cut, copied and pasted between folders

### Changes
* The [app icon](Images/logos/hap-logo.svg) has been updated

### Bug Fixes
* Fixed bug where the app would crash if a user tried to open the upload popover when they had logged out of the app previously, while app restoration made it appear as though they were still logged in
* The title of the current folder is saved when the app is put into the background and closed, so it is shown when the app is resored in the future instead of saying "File Browser"
* When setting the app up for the first time, the school name is now shown correctly on the login screen

## 0.9.0

### Notes
* Starting from build 790, this project uses Xcode 8.2 and Swift 3.0.2. A large amount of code refactoring has taken place since [#f6ecd4] to conform to the latest syntax

### New Stuff
* Log files [can be created](FAQ.md#something-is-not-working-as-expected-using-log-files) for use in debugging the app on remote devices and uploaded
* Photos and videos can be taken and uploaded directly from the app to the current folder, without needing to open another camera app first
* "Pull to refresh" has been enabled for the current browsed-to drive or folder, allowing the listing to be refreshed without having to go up a directory and selecting it again

### Changes
* Added additional supported file icons in the file browser, for file types that are understood by the QuickLook preview

## 0.8.0

### Notes
* This version never left 'alpha' stage, and has since been changed to 0.9.0, due to other commitments causing this project to have an unexpected hiatus (apart from the occasional code fix). The decision to bump up the version number is due to a large number of code changes needed for Swift 3

### New Stuff
* Multiple photos or videos can be selected and uploaded at once to the HAP+ server
* The licenses for the [various projects](README.md#license-terms) this app uses are shown in the main iOS Settings app

### Changes
* The old functionality to upload one photo or video at a time to the HAP+ server can be enabled in the main iOS settings app
* Displayed the name of the file being uploaded on the "file exists" alert, so that it is known which file the message is referring to
* Changed how the "File Browser" table view controller works:
  * If the device is in portrait mode, the view is now shown after login
  * The view can be hidden in both portrait and landscape orentations, giving more available screen space to preview files

### Bug Fixes
* Fixed bug where the app would crash if the user logged out when the device is in portrait mode (The master view controller would stay visible, causing subsequent actions to try and access settings that are nil)
* Fixed bug where the upload popover would show an incorrect lanuch location if the device is rotated when the popover is visible
* Fixed bug where if the currently selected file is downloaded immediately again, but the file data has been changed on a remote device, the correct version of the file is shown

## 0.7.0

### Notes
* Starting from build 729, this project uses Xcode 7.3 and Swift 2.2. These have been updated from Xcode 7.1.1 and Swift 2.1, which were used since this project began

### New Stuff
* Upload files to the HAP+ server from cloud storage providers (the relevant apps need to be installed on the iOS device too)
* Files can be uploaded and downloaded in some external apps without needing to download from the Home Access Plus+ app first
* Users can log out of the app if they want to pass it to another user - [#4](https://github.com/stuajnht/HAP-for-iOS/issues/4)
* Users are automatically logged back in to the HAP+ server and if they do not use the app for a period of time (or their logon sessions are renewed constantly if the app is in the foreground), so they can continue where they left off almost straight away
* Users who are using the device during a timetabled lesson are automatically logged out at the end of the lesson. The device needs to be set up in "shared" mode for this to work, and the [timetable plugin for HAP+](https://hap.codeplex.com/wikipage?title=Timetable%20Plugin&referringTitle=Documentation) needs to be installed

### Changes
* App restoration now takes place if the app is stopped in the background, so that when opened again the user remains in the same folder location they were before
* Updated message on the login screen if the HAP+ server does not have a correctly configured SSL certificate (self signed, expired, domain name mismatch) - credit: [kidpressingbuttons](http://www.edugeek.net/members/kidpressingbuttons.html)

### Bug Fixes
* Fixed typo in upload popover description text from 'curent' to 'current' - [#19](https://github.com/stuajnht/HAP-for-iOS/pull/19) credit: [@TestValleySchool](https://github.com/TestValleySchool)
* Fixed bug where the upload popover would restore incorrectly as a full screen view if the user had to grant app permissions, rather than stay in the popover - [#20](https://github.com/stuajnht/HAP-for-iOS/issues/20)

## 0.6.0

### New Stuff
* Files and folders can be deleted by swiping them to the left
* New folders can be created in the currently viewed folder
* When uploading a file, if it already exists in the current folder the user is given the choice to:
  * Create a new version of the file and keep the old version too
  * Replace the original file with the one being uploaded from the app

### Changes

HomeAccessPlus/FileTableViewCell.swift
// Home Access Plus+ for iOS - A native app to access a HAP+ server
// Copyright (C) 2015-2017  Jonathan Hart (stuajnht) <stuajnht@users.noreply.github.com>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

//
//  FileTableViewCell.swift
//  HomeAccessPlus
//

import UIKit
import Font_Awesome_Swift

class FileTableViewCell: UITableViewCell {

    @IBOutlet weak var lblFileName: UILabel!
    @IBOutlet weak var lblFileType: UILabel!
    @IBOutlet weak var lblFileDetails: UILabel!
    @IBOutlet weak var imgFileIcon: UIImageView!
    
    override func awakeFromNib() {
        super.awakeFromNib()
        // Initialization code
        
        // Setting the colours of the file type and details
        // labels, so that they're not too distracting
        lblFileType.textColor = UIColor.flatGray()
        lblFileDetails.textColor = UIColor.flatGray()
    }

    override func setSelected(_ selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)

        // Configure the view for the selected state
    }
    
    /// Setting the icon for the file or folder, based on what the
    /// extention is of the file type, or a folder if empty
    ///
    /// To help users identify what type of a file is quickly, the
    /// relevant icon should be shown on the table cell, before the
    /// name and description is of the file. The icons used are part
    /// of the FontAwesome font, as file icons do not seem to be
    /// included in the default icons for iOS
    ///
    /// This data is extracted from the JSON response from the HAP+
    /// server when a folder path is specified, and a custom option
    /// is specified here for a network drive.
    ///
    /// Additional file extentions should be added to their relevant
    /// section if they have been missed from this list (which many have!)
    ///
    /// - note: All file types and names are trademarks of their
    ///         respective owners
    ///
    /// - note: Due to the response from the HAP+ API, if a folder contains
    ///         a '.' in it, then it incorrectly adds a folder extension
    ///         into the JSON response, based on whatever is listed after
    ///         the last '.' e.g. 'folder.example' returns '.example' - issue #13
    ///         This may change in a future version of the HAP+ API, so bare
    ///         in mind that someday this function may break!
    ///
    /// - author: Jonathan Hart (stuajnht) <stuajnht@users.noreply.github.com>
    /// - since: 0.3.0-alpha
    /// - version: 6
    /// - date: 2016-04-01
    ///
    /// - parameter fileType: The type of the file of the table cell
    /// - parameter fileExtension: The extension of the file in the table cell
    func fileIcon(_ fileType: String, fileExtension: String) {
        var icon : FAType
        logger.verbose("Setting icon for the file type: \(fileType)")
        
        // Assigning a local variable to hold the file extension, as using
        // 'var' in the function parameter is depricated from Swift 2.2
        // See: https://github.com/apple/swift-evolution/blob/master/proposals/0003-remove-var-parameters.md
        var fileExtension = fileExtension
        
        // Setting the file extension to be "" if the file type is a folder,
        // but contains a '.' in the name somewhere - issue #13
        if (fileType == "Directory") {
            fileExtension = ""
        }
        
        // Setting the file extension to be something custom if the file type
        // is a file, but it contains no extension in the name - issue #13
        let emptyFileExtension = "uk.co.stuajnht.ios.HomeAccessPlus.fileTableViewCell.fileIcon.emptyFileExtension"
        if ((fileType == "File") && (fileExtension == "")) {
            fileExtension = emptyFileExtension
        }
        
        // Seeing what icon should be displayed
        switch fileExtension.lowercased() {
            // Network drive
            case "drive":
                icon = FAType.FAHddO
            
            // File folder
            case "", "directory":
                icon = FAType.FAFolderO
            
            // Adobe Acrobat documents
            case ".pdf":
                icon = FAType.FAFilePdfO
            
            // Archive documents
            case ".7z", ".bz2", ".gz", ".tar", ".zip":
                icon = FAType.FAFileArchiveO
            
            // Audio documents
            case ".aac", ".mp3", ".wav":
                icon = FAType.FAFileAudioO
            
            // Code documents
            case ".c", ".cs", ".css", ".go", ".h", ".html", ".js", ".php", ".pl", ".ps1", ".py", ".sh", ".vb", ".vbs", ".xml":
                icon = FAType.FAFileCodeO
            
            // Image documents
            case ".ai", ".bmp", ".gif", ".icns", ".ico", ".jpeg", ".jpg", ".png", ".psd", ".svg":
                icon = FAType.FAFileImageO
            
            // Microsoft Excel (and similar) documents
            case ".csv", ".ods", ".xls", ".xlsx", ".xlsm":
                icon = FAType.FAFileExcelO
            
            // Microsoft PowerPoint (and similar) documents
            case ".odp", ".ppt", ".pptx", ".pptm":
                icon = FAType.FAFilePowerpointO
            
            // Microsoft Word (and similar) documents
            case ".doc", ".docm", ".docx", ".dotm", ".odt":
                icon = FAType.FAFileWordO
            
            // Text documents
            case ".epub", ".json", ".log", ".rtf", ".txt", ".yml", ".yaml":
                icon = FAType.FAFileTextO
            
            // Video documents
            case ".avi", ".mov", ".mp4", ".mpeg", ".mpg":
                icon = FAType.FAFileVideoO
            
            // Unknown file with no extension - issue #13
            case emptyFileExtension:
                icon = FAType.FAFileO
            
            // Unknown file type
            default:
                icon = FAType.FAFileO
        }
        
        // Displaying the icon
        imgFileIcon.setFAIconWithName(icon: icon, textColor: UIColor(hexString: hapMainColour))
    }

}
